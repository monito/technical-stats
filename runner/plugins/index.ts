export { runPlugins } from './run-plugins'
export { circleci } from './plugin-circleci'
export { docker } from './plugin-docker'
export { github } from './plugin-github'
export { typescript } from './plugin-typescript'
