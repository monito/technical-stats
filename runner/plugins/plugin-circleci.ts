import { gql } from 'graphql-request'
import * as YAML from 'yaml'
import { client } from '../core/api'
import { PluginInput } from '../types'

const QUERY = gql`
  query CircleCi($owner: String!, $name: String!, $configPath: String!) {
    repository(owner: $owner, name: $name) {
      circleciConfig: object(expression: $configPath) {
        ... on Blob {
          text
        }
      }
    }
  }
`

export async function circleci(project: PluginInput) {
  const { owner, name, defaultBranchName } = project
  const { repository } = await client.request(QUERY, {
    owner,
    name,
    configPath: `${defaultBranchName}:.circleci/config.yml`,
  })
  if (!repository.circleciConfig) {
    return {}
  }
  return YAML.parse(repository.circleciConfig.text)
}
