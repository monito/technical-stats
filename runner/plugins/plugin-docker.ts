import { gql } from 'graphql-request'
import { client } from '../core/api'
import { PluginInput } from '../types'

const QUERY = gql`
  query Docker($owner: String!, $name: String!, $dockerfileLocation: String!) {
    repository(owner: $owner, name: $name) {
      dockerfile: object(expression: $dockerfileLocation) {
        ... on Blob {
          text
        }
      }
    }
  }
`

export async function docker (project: PluginInput) {
  const { owner, name, defaultBranchName } = project
  const { repository } = await client.request(QUERY, {
    owner,
    name,
    dockerfileLocation: `${defaultBranchName}:Dockerfile`
  })
  if (!repository.dockerfile) {
    return {}
  }
  const dockerfile = repository.dockerfile.text
  const image: string = dockerfile.split('\n')[0].split('FROM ')[1]
  const versionResults = /.*:(\S+)\s?.*/.exec(image)
  const imageVersion = versionResults && versionResults[1]
  return {
    image,
    imageVersion
  }
}
