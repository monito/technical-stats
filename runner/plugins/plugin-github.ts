import { gql } from 'graphql-request'
import { client } from '../core/api'
import { Project } from '../types'

const QUERY = gql`
  query Github($owner: String!, $name: String!) {
    repository(owner: $owner, name: $name) {
      languages(first: 20, orderBy: { direction: DESC, field: SIZE }) {
        totalSize
        edges {
          size
          node {
            name
          }
        }
      }
    }
  }
`

export async function github (project: Project) {
  const { owner, name } = project
  const { repository } = await client.request(QUERY, { owner, name })
  const { languages: repositoryLanguages } = repository
  const { totalSize } = repositoryLanguages
  const languages = repositoryLanguages.edges
    .map(({ size, node }: { size: number, node: { name: string } }) => ({
      name: node.name,
      percentage: (size / totalSize) * 100,
    }))

  return {
    languages,
  }
}
