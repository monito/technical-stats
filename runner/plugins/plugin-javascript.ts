import { gql } from 'graphql-request'
import { client } from '../core/api'
import { Project } from '../types'

const QUERY = gql`
  query Javascript($owner: String!, $name: String!) {
    repository(owner: $owner, name: $name) {
      package: object(expression: "master:package.json") {
        ... on Blob {
          text
        }
      }
    }
  }
`

export default async function (project: Project) {
  const { owner, name } = project
  const { repository } = await client.request(QUERY, { owner, name })
  const packageJson = repository.package
    ? JSON.parse(repository.package.text)
    : null
  return {
    packageJson,
  }
}
