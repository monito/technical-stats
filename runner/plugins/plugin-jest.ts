import { gql } from 'graphql-request'
import { client } from '../core/api'
import { PluginInput } from '../types'

const QUERY = gql`
  query JestConfig($owner: String!, $name: String!, $configPath: String!) {
    repository(owner: $owner, name: $name) {
      jestConfig: object(expression: $configPath) {
        ... on Blob {
          text
        }
      }
    }
  }
`

export async function jestConfig(project: PluginInput) {
  const { owner, name, defaultBranchName } = project
  let jestConfig
  const { repository } = await client.request(QUERY, {
    owner,
    name,
    configPath: `${defaultBranchName}:jest.config.js`,
  })
  jestConfig = repository.jestConfig?.text
  if (!jestConfig) {
    const { repository } = await client.request(QUERY, {
      owner,
      name,
      configPath: `${defaultBranchName}:jest.config.base.js`,
    })
    jestConfig = repository.jestConfig?.text
    if (!jestConfig) {
      return {}
    }
  }
  return JSON.parse(jestConfig.substring('module.exports = '.length))
}
