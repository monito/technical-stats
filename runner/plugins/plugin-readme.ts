import { gql } from 'graphql-request'
import { client } from '../core/api'
import { PluginInput } from '../types'

const QUERY = gql`
  query readme($owner: String!, $name: String!, $configPath: String!) {
    repository(owner: $owner, name: $name) {
      readme: object(expression: $configPath) {
        ... on Blob {
          text
        }
      }
    }
  }
`

export async function readme(project: PluginInput) {
  const { owner, name, defaultBranchName } = project
  const configPath = `${defaultBranchName}:README.md`
  const {
    repository: { readme },
  } = await client.request(QUERY, { owner, name, configPath })
  return readme?.text
}
