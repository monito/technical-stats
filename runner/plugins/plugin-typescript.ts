import { gql } from 'graphql-request'
import { client } from '../core/api'
import { PluginInput } from '../types'
import * as JSON5 from 'json5'

const QUERY = gql`
  query Typescript($owner: String!, $name: String!, $packageJsonLocation: String!, $tsconfigLocation: String!) {
    repository(owner: $owner, name: $name) {
      package: object(expression: $packageJsonLocation) {
        ... on Blob {
          text
        }
      }
      tsconfig: object(expression: $tsconfigLocation) {
        ... on Blob {
          text
        }
      }
    }
  }
`

export async function typescript (project: PluginInput) {
  const { owner, name, defaultBranchName } = project
  const { repository } = await client.request(QUERY, {
    owner,
    name,
    packageJsonLocation: `${defaultBranchName}:package.json`,
    tsconfigLocation: `${defaultBranchName}:tsconfig.json`
  })

  const tsconfig = repository.tsconfig
    ? JSON5.parse(repository.tsconfig.text)
    : null
  const packageJson = repository.package
    ? JSON.parse(repository.package.text)
    : null
  return { tsconfig, packageJson }
}
