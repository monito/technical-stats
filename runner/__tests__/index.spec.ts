import { run } from '..'
import { github } from '../plugins/plugin-github'
import { typescript } from '../plugins/plugin-typescript'

jest.mock('../core/api')

describe('runner', () => {
  beforeEach(() => {
    console.log = () => {}
  })

  it('can generate report without goals specified', async () => {
    const output = await run({
      organization: 'monito',
      checkAchieved: (percentage: number) => ({
        status: percentage > 75 ? 'pass' : 'fail',
      }),
      plugins: {
        github,
        typescript,
      },
      goals: [],
    })

    expect(output).toMatchSnapshot()
  })

  it('can generate report with goals applied', async () => {
    const output = await run({
      organization: 'monito',
      checkAchieved: (percentage: number) => ({
        status: percentage > 75 ? 'pass' : 'fail',
      }),
      plugins: {
        github,
        typescript,
      },
      goals: [{
        name: 'main branch',
        description: 'the project is using main branch',
        check: async ({ defaultBranchName }) => {
          const status = defaultBranchName === 'main' ? 'pass' : 'fail'
          return { status, value: defaultBranchName }
        }
      }],
    })

    expect(output).toMatchSnapshot()
  })
})
