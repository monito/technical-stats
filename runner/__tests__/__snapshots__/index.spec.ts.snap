// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runner can generate report without goald specified 1`] = `
Object {
  "generatedAt": "2021-05-25T12:24:07.608Z",
  "goals": Array [],
  "organization": "monito",
  "projects": Array [
    Object {
      "achieved": Object {
        "status": "fail",
      },
      "checks": Array [],
      "description": "Tool to generate reports capturing stats of repositories in GitHub organisation ðŸ’š",
      "percentage": 0,
      "repo": "monito/technical-stats",
      "stats": Object {
        "error": 0,
        "fail": 0,
        "pass": 0,
        "skip": 0,
        "warn": 0,
      },
      "url": "https://github.com/monito/technical-stats",
    },
  ],
  "stats": Object {
    "error": 0,
    "fail": 0,
    "pass": 0,
    "skip": 0,
    "warn": 0,
  },
  "totalAchieved": Object {
    "status": "fail",
  },
}
`;

exports[`runner can generate report without goald specified 2`] = `
Object {
  "generatedAt": "2021-05-25T12:24:07.613Z",
  "goals": Array [
    Object {
      "achieved": Object {
        "status": "pass",
      },
      "checks": Array [
        Object {
          "name": "main branch",
          "repo": "monito/technical-stats",
          "status": "pass",
          "value": "main",
        },
      ],
      "description": "the project is using main branch",
      "link": undefined,
      "name": "main branch",
      "percentage": 100,
      "stats": Object {
        "error": 0,
        "fail": 0,
        "pass": 1,
        "skip": 0,
        "warn": 0,
      },
    },
  ],
  "organization": "monito",
  "projects": Array [
    Object {
      "achieved": Object {
        "status": "pass",
      },
      "checks": Array [
        Object {
          "name": "main branch",
          "repo": "monito/technical-stats",
          "status": "pass",
          "value": "main",
        },
      ],
      "description": "Tool to generate reports capturing stats of repositories in GitHub organisation ðŸ’š",
      "percentage": 100,
      "repo": "monito/technical-stats",
      "stats": Object {
        "error": 0,
        "fail": 0,
        "pass": 1,
        "skip": 0,
        "warn": 0,
      },
      "url": "https://github.com/monito/technical-stats",
    },
  ],
  "stats": Object {
    "error": 0,
    "fail": 0,
    "pass": 1,
    "skip": 0,
    "warn": 0,
  },
  "totalAchieved": Object {
    "status": "pass",
  },
}
`;
